Bugzilla/Report/Graph.pm

```diff
@@ -0,0 +1,118 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+#
+# This Source Code Form is "Incompatible With Secondary Licenses", as
+# defined by the Mozilla Public License, v. 2.0.
+
+package Bugzilla::Report::Graph;
+use 5.10.1;
+use Moo;
+
+use Graph::Directed;
+use Graph::Traversal::DFS;
+use Mojo::Util qw(dumper);
+use PerlX::Maybe;
+use Type::Utils     qw(class_type);
+use Types::Standard qw(Bool Int Str ArrayRef Maybe);
+use Set::Object     qw(set);
+
+use Bugzilla;
+use Bugzilla::Logging;
+use Bugzilla::Types qw(DB);
+
+has bug_id => (is => 'ro',   isa => Int, required => 1);
+has table  => (is => 'ro',   isa => Str, default  => 'dependencies');
+has depth  => (is => 'ro',   isa => Int, default  => 3);
+has source => (is => 'ro',   isa => Str, default  => 'dependson');
+has sink   => (is => 'ro',   isa => Str, default  => 'blocked');
+has limit  => (is => 'ro',   isa => Maybe [Int], default => 10_000);
+has paths  => (is => 'lazy', isa => ArrayRef [ArrayRef]);
+has graph  => (is => 'lazy', isa => class_type({class => 'Graph'}));
+has query  => (is => 'lazy', isa => Str);
+
+# Run the query that will list of the paths from the parent bug
+# down to the last child in the tree
+sub _build_paths {
+  my ($self) = @_;
+  return Bugzilla->dbh->selectall_arrayref($self->query, undef, $self->bug_id);
+}
+
+# Builds a new directed graph
+sub _build_graph {
+  my ($self) = @_;
+  my $paths  = $self->paths;
+  my $graph  = Graph::Directed->new;
+
+  foreach my $path (@$paths) {
+    pop @$path until defined $path->[-2];
+    $graph->add_path(@$path);
+  }
+
+  return $graph;
+}
+
+sub _build_query {
+  my ($self) = @_;
+  my $table  = $self->table;
+  my $alias  = substr $table, 0, 1;
+  my $depth  = $self->depth;
+  my $source = $self->source;
+  my $sink   = $self->sink;
+  my $limit  = defined $self->limit ? 'LIMIT ' . $self->limit : '';
+
+  # WITH RECURSIVE is available in MySQL 8.x and newer as
+  # well as recent versions of PostgreSQL and SQLite.
+  my $query = "WITH RECURSIVE RelationshipTree AS (
+    SELECT t.$source, t.$sink, 1 AS depth FROM $table t WHERE t.$source = ?
+    UNION ALL
+    SELECT t.$source, t.$sink, rt.depth + 1 AS depth FROM $table t
+      JOIN RelationshipTree rt ON t.$source = rt.$sink WHERE rt.depth <= $depth $limit)
+    SELECT rt.$source, rt.$sink FROM RelationshipTree rt";
```

Comment:
Can we either build this query with a pre-existing query builder library? This seems like a target for a SQL injection given we are splatting strings directly into the query. If source/sink etc are all pre-defined values perhaps we could at least validate them here.